docs/blueprints/referral-manager-v1.md

```diff
@@ -0,0 +1,206 @@
+    
+## Rationale
+
+We will create an admin port that's only listening on `localhost` so CLI could be only used by Referral Manager operators.
+
+We could also set the admin port to only accept requests coming from storj vpn.
+
+## Implementation
+
+- Create a private repository for Referral Manager.
+- Create `tokens` and `users` table in Referral Manager database.
+- Create an admin endpoint on Referral Manager.
+    - Implementing a method `GenerateTokens` for upadting the `users` table on Referral Manager to contain new tokens for eligible users.
+- Implementing an endpoint `GetTokens` for requesting unredeemed tokens from Referral Manager. 
+- Implementing an endpoint `Redeem` on Referral Manager for verifying invitation tokens and storing newly created user ID into `users` table.
+- Replace existing registration token logic.
+
+### Pseudocode
+
+The Token struct represents both a Go struct (on satellite and Referral Manager) as well as the schema for the `tokens` table on the Referral Manager
+
+```
+type Token struct {
+    Secret [32]byte
+    OwnerID uuid.UUID
+    RedeemedID uuid.UUID
+    OwnerSatelliteURL string
+    RedeemedSatelliteURL string
+    Status string
+}
+```
+Statuses: `unsent` (owner's satellite doesn't know about it yet), `unredeemed` (owner's satellite knows about it but it has not been used), `redeemed` (someone has used this referral link to register alreadys)
+
+The User struct represent both a Go struct as well as the schema for the `users` table on Referral Manager.
+
+```
+type User struct {
+    ID uuid.UUID
+    satelliteURL string
+    NewTokens int
+    UnredeemedTokens int
+}
+```
+
+
+**Endpoints for satellites:**
+
+```
+// GetTokens retrieves a list of unredeemed tokens for a user.
+GetTokens(userID uuid.UUID, satelliteURL string) []tokens {
+    user, err := db.GetUser(userID, satelliteURL)
+    if err == UserNotFound {
+        db.CreateUser(userID, satelliteURL)
+        return nil
+    }
+    if user.NewTokens > 0 {
+        for i:=0; i<user.NewTokens; i++ {
+            newToken := Token{data: generateRandomToken(), user: user.ID, satellite: user.satelliteURL}
+            db.CreateToken(newToken)
+            tokens = append(tokens, newToken)
+        }
+        db.UpdateUserNewTokens(userID, satelliteURL, 0)
+    }
+    tokens := db.GetTokensByUserIDAndSatelliteURL(userID, satelliteURL)
+
+    return tokens
+}
+
+// Redeem marks a token as redeemed and stores user info into database
+func Redeem(ctx, token, userID, satelliteURL) error {
+	// only update the status if the status of a token is unredeems
+	tokenStatus, err := db.RedeemToken(ctx, token, userID)
+	if err != nil {
+		return err
+	}
+
+    // decrease unredeemed token count in users table by 1
+    err := db.UpdateUserInfo(ctx, user)
+    if err != nil {
+        return err
+    }
+
+    // save user info into users table
+    err := db.CreateUser(ctx, userID, satelliteURL)
```

Comment:
If we do this in a transaction with the other parts, then we would get an error if the userID already exists and the above changes would be rolled back. Just in case something goes wrong somewhere.